cunumeric/deferred.py

```diff
@@ -3487,25 +3487,68 @@ def scan(
             assert self.shape == swapped.shape
             self.copy(swapped, deep=True)
 
-    def unique(self) -> NumPyThunk:
-        result = self.runtime.create_unbound_thunk(self.base.type)
-
+    def unique(
+        self, return_index: bool = False
+    ) -> Union[NumPyThunk, tuple[NumPyThunk, Optional[NumPyThunk]]]:
         task = self.context.create_auto_task(CuNumericOpCode.UNIQUE)
 
-        task.add_output(result.base)
         task.add_input(self.base)
+        task.add_scalar_arg(return_index, ty.bool_)
 
+        result = None
+        # Assuming legate core will always choose GPU variant
         if self.runtime.num_gpus > 0:
             task.add_nccl_communicator()
+            result = self.runtime.create_unbound_thunk(self.base.type)
+        elif return_index:
+            result = self.runtime.create_unbound_thunk(
+                ty.struct_type(
+                    [
+                        self.base.type,
+                        ty.int64,
+                    ],
+                    True,
+                )
+            )
+        else:
+            result = self.runtime.create_unbound_thunk(self.base.type)
+        task.add_output(result.base)
+
+        returned_indices = None
+        if return_index:
+            returned_indices = self.runtime.create_unbound_thunk(ty.int64)
+            if self.runtime.num_gpus > 0:
+                task.add_output(returned_indices.base)
+
+            for i in range(self.ndim):
+                task.add_scalar_arg(self.shape[i], ty.int32)
 
         task.execute()
 
-        if self.runtime.num_gpus == 0 and self.runtime.num_procs > 1:
-            result.base = self.context.tree_reduce(
-                CuNumericOpCode.UNIQUE_REDUCE, result.base
-            )
+        if self.runtime.num_gpus == 0:
+            if self.runtime.num_procs > 1:
+                result.base = self.context.tree_reduce(
+                    CuNumericOpCode.UNIQUE_REDUCE,
+                    result.base,
+                    scalar_args=[(return_index, ty.bool_)],
+                )
+            if return_index:
+                task = self.context.create_auto_task(CuNumericOpCode.UNZIP)
+                task.add_input(result.base)
```

Comment:
this is an unclear way of writing the code, adding `result.base` as input, overwriting result, and then adding `result.base` as output.